src/main/java/org/springframework/data/r2dbc/dialect/NamedBindMarkers.java

```diff
@@ -0,0 +1,129 @@
+package org.springframework.data.r2dbc.dialect;
+
+import io.r2dbc.spi.Statement;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import org.springframework.util.Assert;
+
+/**
+ * Name-based bind markers.
+ *
+ * @author Mark Paluch
+ */
+class NamedBindMarkers implements BindMarkers {
+
+	private static final AtomicIntegerFieldUpdater<NamedBindMarkers> COUNTER_INCREMENTER = AtomicIntegerFieldUpdater
+			.newUpdater(NamedBindMarkers.class, "counter");
+
+	// access via COUNTER_INCREMENTER
+	@SuppressWarnings("unused") private volatile int counter;
+
+	private final String prefix;
+
+	private final String indexPrefix;
+
+	private final int nameLimit;
+
+	NamedBindMarkers(String prefix, String indexPrefix, int nameLimit) {
+
+		this.prefix = prefix;
+		this.indexPrefix = indexPrefix;
+		this.nameLimit = nameLimit;
+	}
+
+	@Override
+	public BindMarker next() {
+
+		String name = nextName();
+
+		return new NamedBindMarker(prefix + name, name);
+	}
+
+	@Override
+	public BindMarker next(String nameHint) {
+
+		Assert.notNull(nameHint, "Name hint must not be null");
+
+		String name = nextName();
+
+		String filteredNameHint = filter(nameHint);
+
+		if (!filteredNameHint.isEmpty()) {
+			name += "_" + filteredNameHint;
```

Comment:
I find it slightly inconsistent that the `indexPrefix` is configurable but the `nameHintPrefix` isn't.

Reply:
I think it makes sense to move the actual filtering and concatenation into a `Dialect`-specific component and reduce named bind markers to their essence.